yt_experiments/octree/converter.pyx

```diff
@@ -0,0 +1,281 @@
+        unique_index : int
+            A unique index for the cell.
+
+        Returns
+        -------
+        Oct*
+            The node in the tree at the cell location.
+        """
+        cdef Oct* node = self.add(x, level, unique_index)
+
+        if (
+            not np.isclose(node.x, x[0]) or
+            not np.isclose(node.y, x[1]) or
+            not np.isclose(node.z, x[2])
+        ):
+            raise ValueError(
+                "Node xc does not match. Expected "
+                f"{x[0]}, {x[1]}, {x[2]}, got {node.x}, {node.y}, {node.z} @ level {level}"
+            )
+
+        if node.ind != unique_index:
+            raise ValueError(
+                "Node ind already set. Make sure that the same cell is not added twice."
+            )
+
+        return node
+
+    @cython.boundscheck(False)
+    cdef Oct* add(self, const double[3] x, const int level, const int unique_index) noexcept:
+        """Add a cell to the octree.
+
+        Parameters
+        ----------
+        x : double[3]
+            The position of the cell in unitary units (within [0, 1])
+        level : int
+            The level of the cell. It should be such that x * 2**level is an integer
+        unique_index : int
+            A unique index for the cell.
+
+        Returns
+        -------
+        Oct*
+            The node in the tree at the cell location.
+        """
+        cdef float dx = 0.5
+
+        cdef size_t ilvl, i, ind
+        cdef np.int8_t ix, iy, iz
+
+        cdef Oct* node = self.root
+        cdef Oct* child = self.root
+
+        for ilvl in range(level):
+            ix = x[0] > node.x
+            iy = x[1] > node.y
+            iz = x[2] > node.z
+            ind = 4 * ix + 2 * iy + iz
+
+            if node.children == NULL:
+                node.children = <Oct **> malloc(8 * sizeof(Oct*))
+                for i in range(8):
+                    node.children[i] = NULL
+
+            if node.children[ind] == NULL:
+                # Create a new node
+                child = <Oct*>malloc(sizeof(Oct))
+                child.children = NULL
+                child.x = node.x + 0.5 * dx * (2 * ix - 1)
+                child.y = node.y + 0.5 * dx * (2 * iy - 1)
+                child.z = node.z + 0.5 * dx * (2 * iz - 1)
+                child.ind = -1
+
+                self.count += 1
+
+                node.children[ind] = child
+
+            dx *= 0.5
+
+            node = node.children[ind]
+
+        if node.ind == -1:
+            node.ind = unique_index
+        return node
+
+    @classmethod
+    @cython.boundscheck(False)
+    def from_list(cls, const cython.floating[:, ::1] Xc, const cython.integral[::1] levels, bint check = False):
+        """Create an octree from a list of cells.
+
+        Parameters
+        ----------
+        Xc : double array (N, 3)
+            The positions of the cells in unitary units (within [0, 1])
+        levels : int array (N,)
```

Comment:
i'm confused by this description (and in the other function docstrings) -- I don't see how `Xc * 2**level` will always be an integer? e.g., with the example:

```python
>>> xyz = np.array(
...             [
...                 [0.375, 0.125, 0.125],
...                 [0.125, 0.125, 0.125],
...                 [0.375, 0.375, 0.375],
...                 [0.75, 0.75, 0.75],
...             ])
>>> levels = np.array([2, 2, 2, 1], dtype=np.int32)
>>> for i in range(0,len(levels)):
...     print(xyz[i,:] * 2**levels[i])
... 
[1.5 0.5 0.5]
[0.5 0.5 0.5]
[1.5 1.5 1.5]
[1.5 1.5 1.5]
```

Reply:
My mistake, it should be half-integers. I've fixed the docstring now :)